Dynamic Screen Rotation

ImpulseAdventure opened the pull request:
This PR merges the contributions from @espHorst PR #70 along with some additional edits.
Please refer to #70 for details.

---
**Add GuiRotate() to support dynamic screen rotation (#70) in DRV_DISP_ADAGFX**

- NOTE: Breaking change for configuration update. GUIslice_config_*.h will
  require an update to add the new GSLC_TOUCH_ROTATE parameter. The most
  minimal update would require adding the following line:
  `#define GSLC_TOUCH_ROTATE 1` (the number may need to be changed to a
  value between 0..3 to provide correct touch detection)
- Add GuiRotate() as simple wrapper for DrvRotate(). Note that only
  DRV_DISP_ADAGFX and DRV_DISP_ADAGFX_AS modes are supported at the moment,
  with others to follow.
- Add compiler warning to detect config file needing update for GSLC_TOUCH_ROTATE
- Relocated TOUCH_ROTATION_* macros into GUIslice.h from config
- Update example gslc_ex17_ard to use GuiRotate instead of DrvRotate

espHorst commented:
Changing the default value in the case of an undefined GSLC_TOUCH_ROTATE (see comment in code) I think is fully backwards compatible. Thus maybe it isnt a 
>NOTE: Breaking change for configuration update. GUIslice_config_*.h will
require an update 
breaking change?

ImpulseAdventure commented:
Even though isn't a breaking change from the perspective of halting compilation, I listed it as breaking because one's original (unmodified) configuration may now give incorrect touch behavior.

For example, until now, the library's default config_ard lists:
```
  #define GSLC_ROTATE     1
  #define ADATOUCH_SWAP_XY  1
  #define ADATOUCH_FLIP_X   0
  #define ADATOUCH_FLIP_Y   1
```
If left as-is, the user will be notified with the preprocessor directive, and then `GSLC_TOUCH_ROTATE` will be assigned (in GUIslice.h) to be equal to `GSLC_ROTATE` (which itself is still 1). Therefore, I believe the new combination of (default) GSLC_TOUCH_ROTATE and the old `ADATOUCH_*` settings will lead to differing touch handling than they had before this update.

To be on the safe said, I have regarded this update as breaking due to the fact that I think an unedited config file may now lead to incorrect touch behavior.

Might there be a better way to default this?

espHorst commented:
As long as you define `#define GSLC_TOUCH_ROTATE 0` the old behaviour is not changed at all.
GSLC_TOUCH_ROTATE == 0 does not change the rotation.
In this case the values defined in GUIslice_config_ard.h, e.g. 
```
  #define GSLC_ROTATE     1
  #define ADATOUCH_SWAP_XY  1
  #define ADATOUCH_FLIP_X   0
  #define ADATOUCH_FLIP_Y   1
```
are directly used. Thus I propose to define  `#define GSLC_TOUCH_ROTATE 0` in the case that GSLC_TOUCH_ROTATE is not defined in the GUIslice_config_ard.h file.
An unedited config file will thus work as before. 
Thus I propose to change your GUIslice.h proposal to
```
#ifndef GSLC_TOUCH_ROTATE
  // Apply a default which does not change anything
  #define GSLC_TOUCH_ROTATE 0
#endif
```
User just can keep their  working ADATOUCH_* & GSLC_ROTATE configuration. Dynamic rotation will work nevertheless (if I made no mistake). 

Summary:

GSLC_TOUCH_ROTATE and the old style way using ADATOUCH_SWAP_XY, ADATOUCH_FLIP_X,ADATOUCH_FLIP_Y can be mixed. Defining `#define GSLC_TOUCH_ROTATE 0` will keep the existing behaviour. Dynamic rotation will work in any case (if I made no mistake). 
In the current version of GUIslice_config_ard.h we introduce `#define GSLC_TOUCH_ROTATE 1` and set the swap/flip defines to 0. If users use their old GUIslice_config_ard.h then they will still work as
GSLC_TOUCH_ROTATE is not defined and therefore it is set to 0 by the code in GUIslice.h (see code  above).

ImpulseAdventure commented:
Good point, you're right. Specifying 0 for the default should enable full backwards-compatibility. The net result is that this PR should not be a breaking change.

I have revised the PR commits, and also modified the description in the config file a bit (Note that I removed the text associated with `GSLC_TOUCH_ROTATE_OFFSET` as the code doesn't exist for that yet).

I'll run my regression again and merge the PR if it continues to look OK.

thanks!
